File: test-mimo-api.py
"""
MiMo API 完整测试脚本 (Python版本)
严格按照源码实现的请求格式
"""

import requests
import json
import uuid
from typing import Optional, Dict, Any

# ==================== 配置区域 ====================
# 请将从浏览器获取的信息填写在这里
CONFIG = {
    # 从Cookie中获取的三个关键值
    'userId': 'YOUR_USER_ID_HERE',           # 例如: '123456789'
    'xiaomichatbot_ph': 'YOUR_PH_HERE',      # 例如: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...'
    'serviceToken': 'YOUR_SERVICE_TOKEN_HERE', # 例如: 'abcdef123456789'
    
    # 模型配置
    'model': 'mimo-v2.5',  # 可选: mimo-v2.5, mimo-v2.5-pro, mimo-v2-flash
    'temperature': 0.7,
    'topP': 0.9,
    
    # 是否启用思考模式
    'enableThinking': False
}

# ==================== 工具函数 ====================

def generate_uuid() -> str:
    """生成UUID（无连字符）"""
    return str(uuid.uuid4()).replace('-', '')

def build_cookie() -> str:
    """构建Cookie字符串"""
    return f"userId={CONFIG['userId']}; xiaomichatbot_ph={CONFIG['xiaomichatbot_ph']}; serviceToken={CONFIG['serviceToken']}"

def build_headers() -> Dict[str, str]:
    """构建请求头（完全按照源码）"""
    return {
        'Content-Type': 'application/json',
        'Accept': 'text/event-stream, text/plain, */*',
        'Cookie': build_cookie(),
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/120.0.0.0',
        'Referer': 'https://aistudio.xiaomimimo.com/',
        'Origin': 'https://aistudio.xiaomimimo.com'
    }

def build_payload(query: str, conversation_id: Optional[str] = None, parent_id: Optional[str] = None) -> Dict[str, Any]:
    """构建请求体（完全按照源码）"""
    # 处理模型名称（flash模型需要添加-studio后缀）
    model_name = CONFIG['model'].lower()
    if 'flash' in model_name and not model_name.endswith('-studio'):
        model_name = f"{model_name}-studio"
    
    return {
        'msgId': generate_uuid(),
        'conversationId': conversation_id or generate_uuid()[:32],
        'query': query,
        'messages': [],
        'parentId': parent_id,
        'save': True,
        'source': 'STATION',
        'scene': 'STATION',
        'modelConfig': {
            'enableThinking': CONFIG['enableThinking'],
            'model': model_name,
            'temperature': CONFIG['temperature'],
            'topP': CONFIG['topP']
        },
        'multiMedias': []
    }

# ==================== API 调用 ====================

def send_chat_request(query: str, conversation_id: Optional[str] = None, parent_id: Optional[str] = None) -> Dict[str, Any]:
    """发送聊天请求"""
    url = f"https://aistudio.xiaomimimo.com/open-apis/bot/chat?xiaomichatbot_ph={requests.utils.quote(CONFIG['xiaomichatbot_ph'])}"
    headers = build_headers()
    payload = build_payload(query, conversation_id, parent_id)
    
    print('\n' + '='*60)
    print('🚀 发送请求到 MiMo API')
    print('='*60)
    print(f'URL: {url}')
    print('\n📋 请求头:')
    print(json.dumps(headers, indent=2, ensure_ascii=False))
    print('\n📦 请求体:')
    print(json.dumps(payload, indent=2, ensure_ascii=False))
    print('\n' + '='*60)
    
    try:
        response = requests.post(
            url,
            headers=headers,
            json=payload,
            timeout=60,
            stream=True
        )
        
        print(f'\n✅ 请求成功！')
        print(f'状态码: {response.status_code}')
        print(f'响应头: {json.dumps(dict(response.headers), indent=2, ensure_ascii=False)}')
        print('\n' + '='*60)
        print('📥 响应内容（流式）:')
        print('='*60)
        
        full_response = ''
        last_msg_id = '0'
        returned_conv_id = payload['conversationId']
        
        for line in response.iter_lines():
            if not line:
                continue
            
            line_str = line.decode('utf-8')
            
            if line_str.startswith('id:'):
                last_msg_id = line_str.replace('id:', '').strip()
            
            if line_str.startswith('data:'):
                try:
                    data_str = line_str.replace('data:', '').strip()
                    if data_str == '[DONE]':
                        print('\n[完成]')
                        continue
                    
                    data = json.loads(data_str)
                    
                    # 提取对话ID
                    if 'conversationId' in data:
                        returned_conv_id = data['conversationId']
                    
                    # 提取消息ID
                    if 'id' in data:
                        last_msg_id = data['id']
                    
                    # 提取回复内容
                    if 'text' in data:
                        print(data['text'], end='', flush=True)
                        full_response += data['text']
                    
                    # 提取思考内容（如果有）
                    if 'thinking' in data:
                        print('\n[思考中...]')
                        print(data['thinking'], end='', flush=True)
                
                except json.JSONDecodeError:
                    pass
        
        print('\n\n' + '='*60)
        print('🎉 响应完成！')
        print('='*60)
        print(f'会话ID: {returned_conv_id}')
        print(f'消息ID: {last_msg_id}')
        print(f'完整回复长度: {len(full_response)} 字符')
        
        return {
            'conversationId': returned_conv_id,
            'msgId': last_msg_id,
            'response': full_response
        }
    
    except requests.exceptions.RequestException as error:
        print(f'\n❌ 请求失败！')
        print(f'错误信息: {error}')
        
        if hasattr(error, 'response') and error.response is not None:
            print(f'状态码: {error.response.status_code}')
            print(f'响应数据: {error.response.text}')
        
        raise error

def validate_config() -> bool:
    """验证配置"""
    print('🔍 验证配置...')
    
    if (CONFIG['userId'] == 'YOUR_USER_ID_HERE' or
        CONFIG['xiaomichatbot_ph'] == 'YOUR_PH_HERE' or
        CONFIG['serviceToken'] == 'YOUR_SERVICE_TOKEN_HERE'):
        print('\n❌ 错误：请先在CONFIG中填写从浏览器获取的信息！')
        print('\n请参考 MiMo-测试指南.md 获取以下信息：')
        print('  - userId')
        print('  - xiaomichatbot_ph')
        print('  - serviceToken')
        return False
    
    print('✅ 配置验证通过')
    print(f"  - userId: {CONFIG['userId']}")
    print(f"  - xiaomichatbot_ph: {CONFIG['xiaomichatbot_ph'][:50]}...")
    print(f"  - serviceToken: {CONFIG['serviceToken']}")
    print(f"  - model: {CONFIG['model']}")
    
    return True

# ==================== 主函数 ====================

def main():
    print('\n' + '='*60)
    print('MiMo API 测试')
    print('='*60)
    
    # 验证配置
    if not validate_config():
        return
    
    try:
        # 测试1：简单对话
        print('\n📝 测试1：简单对话')
        result1 = send_chat_request('你好，请用一句话介绍你自己')
        
        # 测试2：多轮对话（使用相同的会话ID）
        print('\n\n📝 测试2：多轮对话')
        result2 = send_chat_request(
            '刚才我说了什么？',
            result1['conversationId'],
            result1['msgId']
        )
        
        print('\n\n' + '='*60)
        print('✨ 所有测试完成！')
        print('='*60)
    
    except Exception as error:
        print(f'\n\n💥 测试失败: {error}')
        print('\n可能的原因：')
        print('  1. Cookie已过期，请重新获取')
        print('  2. 网络连接问题')
        print('  3. 请求参数错误')

if __name__ == '__main__':
    main()

File: test-mimo-api.js
/**
 * MiMo API 完整测试脚本
 * 严格按照源码实现的请求格式
 */

const axios = require('axios');

// ==================== 配置区域 ====================
// 请将从浏览器获取的信息填写在这里
const CONFIG = {
    // 从Cookie中获取的三个关键值
    userId: '2937012631',
    xiaomichatbot_ph: 'b109A8ByzzXVgKrToJ8uFQ==',
    serviceToken: '/Qzv9hyEQZikhu2h7j2kvFiFJIoPjgLHzdhQAm7xyQkKnH4D4g28Xac4J2LplmjNZTG7fD/8BB7TojhGP51gcsiw64Nl2EhhQYdJ1MuOL0rPaCJ85lvWDhCHMh2hqDgNV1BKI0lyr/4qTz+tLYowN3c/APlMlk0jPgJLaIapOZhbFj7tqFPhWqLGjp/aGuImA2UEc29H8p3onlZewLIeQOYjGkkZ79AMN0RY9XdsnZAxvVjhnIok9NIdZiUGsO7p/LQ/DNc5ZgKbQs5LLDHLOOFdB0HJtsibs8o9ZOBb0lkoRZJXo6wCBKkLIGoEQUhhO4i728TMf/DIyfgzrCeuOjH4QUt8W4FLIBolnJUNINQ=',
    
    // 模型配置
    model: 'mimo-v2.5',  // 可选: mimo-v2.5, mimo-v2.5-pro, mimo-v2-flash
    temperature: 0.7,
    topP: 0.9,
    
    // 是否启用思考模式
    enableThinking: false
};

// ==================== 工具函数 ====================

/**
 * 生成UUID（无连字符）
 */
function generateUUID() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
        const r = Math.random() * 16 | 0;
        const v = c === 'x' ? r : (r & 0x3 | 0x8);
        return v.toString(16);
    }).replace(/-/g, ''); // 移除连字符
}

/**
 * 构建Cookie字符串
 */
function buildCookie() {
    return `userId=${CONFIG.userId}; xiaomichatbot_ph=${CONFIG.xiaomichatbot_ph}; serviceToken=${CONFIG.serviceToken}`;
}

/**
 * 构建请求头（完全按照源码）
 */
function buildHeaders() {
    return {
        'Content-Type': 'application/json',
        'Accept': 'text/event-stream, text/plain, */*',
        'Cookie': buildCookie(),
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/120.0.0.0',
        'Referer': 'https://aistudio.xiaomimimo.com/',
        'Origin': 'https://aistudio.xiaomimimo.com'
    };
}

/**
 * 构建请求体（完全按照源码）
 */
function buildPayload(query, conversationId = null, parentId = null) {
    // 处理模型名称（flash模型需要添加-studio后缀）
    let modelName = CONFIG.model.toLowerCase();
    if (modelName.includes('flash') && !modelName.endsWith('-studio')) {
        modelName = `${modelName}-studio`;
    }
    
    return {
        msgId: generateUUID(),
        conversationId: conversationId || generateUUID().substring(0, 32),
        query: query,
        messages: [],
        parentId: parentId,
        save: true,
        source: 'STATION',
        scene: 'STATION',
        modelConfig: {
            enableThinking: CONFIG.enableThinking,
            model: modelName,
            temperature: CONFIG.temperature,
            topP: CONFIG.topP
        },
        multiMedias: []
    };
}

// ==================== API 调用 ====================

/**
 * 发送聊天请求
 */
async function sendChatRequest(query, conversationId = null, parentId = null) {
    const url = `https://aistudio.xiaomimimo.com/open-apis/bot/chat?xiaomichatbot_ph=${encodeURIComponent(CONFIG.xiaomichatbot_ph)}`;
    const headers = buildHeaders();
    const payload = buildPayload(query, conversationId, parentId);
    
    console.log('\n' + '='.repeat(60));
    console.log('🚀 发送请求到 MiMo API');
    console.log('='.repeat(60));
    console.log('URL:', url);
    console.log('\n📋 请求头:');
    console.log(JSON.stringify(headers, null, 2));
    console.log('\n📦 请求体:');
    console.log(JSON.stringify(payload, null, 2));
    console.log('\n' + '='.repeat(60));
    
    try {
        const response = await axios.post(url, payload, {
            headers: headers,
            timeout: 60000,
            responseType: 'stream'
        });
        
        console.log('\n✅ 请求成功！');
        console.log('状态码:', response.status);
        console.log('响应头:', JSON.stringify(response.headers, null, 2));
        console.log('\n' + '='.repeat(60));
        console.log('📥 响应内容（流式）:');
        console.log('='.repeat(60));
        
        let fullResponse = '';
        let lastMsgId = '0';
        let returnedConvId = payload.conversationId;
        
        return new Promise((resolve, reject) => {
            response.data.on('data', (chunk) => {
                const chunkStr = chunk.toString('utf8');
                
                const lines = chunkStr.split('\n');
                for (const line of lines) {
                    if (line.startsWith('id:')) {
                        lastMsgId = line.replace('id:', '').trim();
                    }
                    if (line.startsWith('event:')) {
                        // 事件类型，可以忽略
                    }
                    if (line.startsWith('data:')) {
                        try {
                            const dataStr = line.replace('data:', '').trim();
                            if (dataStr === '[DONE]' || !dataStr) {
                                continue;
                            }
                            const data = JSON.parse(dataStr);
                            
                            // 提取对话ID
                            if (data.conversationId) {
                                returnedConvId = data.conversationId;
                            }
                            
                            // 提取消息ID
                            if (data.id) {
                                lastMsgId = data.id;
                            }
                            
                            // 提取回复内容 (注意：字段名是content，不是text)
                            if (data.content) {
                                process.stdout.write(data.content);
                                fullResponse += data.content;
                            }
                            
                            // 提取思考内容（如果有）
                            if (data.thinking) {
                                console.log('\n[思考中...]');
                                process.stdout.write(data.thinking);
                            }
                            
                        } catch (e) {
                            // 忽略解析错误
                        }
                    }
                }
            });
            
            response.data.on('end', () => {
                console.log('\n\n' + '='.repeat(60));
                console.log('🎉 响应完成！');
                console.log('='.repeat(60));
                console.log('会话ID:', returnedConvId);
                console.log('消息ID:', lastMsgId);
                console.log('完整回复长度:', fullResponse.length, '字符');
                
                resolve({
                    conversationId: returnedConvId,
                    msgId: lastMsgId,
                    response: fullResponse
                });
            });
            
            response.data.on('error', (error) => {
                console.error('\n❌ 流式响应错误:', error);
                reject(error);
            });
        });
        
    } catch (error) {
        console.error('\n❌ 请求失败！');
        console.error('错误信息:', error.message);
        
        if (error.response) {
            console.error('状态码:', error.response.status);
            console.error('响应数据:', error.response.data);
        }
        
        throw error;
    }
}

/**
 * 验证配置
 */
function validateConfig() {
    console.log('🔍 验证配置...');
    
    if (CONFIG.userId === 'YOUR_USER_ID_HERE' ||
        CONFIG.xiaomichatbot_ph === 'YOUR_PH_HERE' ||
        CONFIG.serviceToken === 'YOUR_SERVICE_TOKEN_HERE') {
        console.error('\n❌ 错误：请先在CONFIG中填写从浏览器获取的信息！');
        console.log('\n请参考 MiMo-测试指南.md 获取以下信息：');
        console.log('  - userId');
        console.log('  - xiaomichatbot_ph');
        console.log('  - serviceToken');
        return false;
    }
    
    console.log('✅ 配置验证通过');
    console.log('  - userId:', CONFIG.userId);
    console.log('  - xiaomichatbot_ph:', CONFIG.xiaomichatbot_ph.substring(0, 50) + '...');
    console.log('  - serviceToken:', CONFIG.serviceToken);
    console.log('  - model:', CONFIG.model);
    
    return true;
}

// ==================== 主函数 ====================

async function main() {
    console.log('\n' + '='.repeat(60));
    console.log('MiMo API 测试');
    console.log('='.repeat(60));
    
    // 验证配置
    if (!validateConfig()) {
        return;
    }
    
    try {
        // 测试1：简单对话
        console.log('\n📝 测试1：简单对话');
        const result1 = await sendChatRequest('你好，请用一句话介绍你自己');
        
        // 测试2：多轮对话（使用相同的会话ID）
        console.log('\n\n📝 测试2：多轮对话');
        const result2 = await sendChatRequest(
            '刚才我说了什么？',
            result1.conversationId,
            result1.msgId
        );
        
        console.log('\n\n' + '='.repeat(60));
        console.log('✨ 所有测试完成！');
        console.log('='.repeat(60));
        
    } catch (error) {
        console.error('\n\n💥 测试失败:', error.message);
        console.log('\n可能的原因：');
        console.log('  1. Cookie已过期，请重新获取');
        console.log('  2. 网络连接问题');
        console.log('  3. 请求参数错误');
    }
}

// 执行测试
main();
